src/include/api.h

```diff
@@ -167,24 +169,24 @@
 #define SESSION_API_CALL_PREPARE_ALLOWED(s, n, config, cfg) \
     API_CALL(s, WT_SESSION, n, NULL, config, cfg)
 
-#define SESSION_API_CALL(s, n, config, cfg)        \
-    API_CALL(s, WT_SESSION, n, NULL, config, cfg); \
-    WT_ERR(__wt_txn_context_prepare_check((s)))
+#define SESSION_API_CALL(s, n, config, cfg)                        \
+    WT_RET(__wt_txn_context_prepare_check((s), "WT_SESSION." #n)); \
+    API_CALL(s, WT_SESSION, n, NULL, config, cfg)
 
 #define SESSION_API_CALL_NOCONF(s, n) API_CALL_NOCONF(s, WT_SESSION, n, NULL)
 
-#define SESSION_API_CALL_NOCONF_PREPARE_NOT_ALLOWED(s, n) \
-    API_CALL_NOCONF(s, WT_SESSION, n, NULL);              \
-    WT_ERR(__wt_txn_context_prepare_check((s)))
+#define SESSION_API_CALL_NOCONF_PREPARE_NOT_ALLOWED(s, n)          \
+    WT_RET(__wt_txn_context_prepare_check((s), "WT_SESSION." #n)); \
+    API_CALL_NOCONF(s, WT_SESSION, n, NULL)
 
-#define SESSION_TXN_API_CALL(s, n, config, cfg)        \
-    TXN_API_CALL(s, WT_SESSION, n, NULL, config, cfg); \
-    WT_ERR(__wt_txn_context_prepare_check((s)))
+#define SESSION_TXN_API_CALL(s, n, config, cfg)                    \
+    WT_RET(__wt_txn_context_prepare_check((s), "WT_SESSION." #n)); \
+    TXN_API_CALL(s, WT_SESSION, n, NULL, config, cfg)
 
 #define CURSOR_API_CALL(cur, s, n, bt)                                                     \
     (s) = (WT_SESSION_IMPL *)(cur)->session;                                               \
+    WT_RET(__wt_txn_context_prepare_check((s), "WT_CURSOR." #n));                          \
     API_CALL_NOCONF(s, WT_CURSOR, n, ((bt) == NULL) ? NULL : ((WT_BTREE *)(bt))->dhandle); \
-    WT_ERR(__wt_txn_context_prepare_check((s)));                                           \
```

Comment:
So has this been wrong the entire time? I.e. we used to check after the `API_CALL` macro? Or is this necessary based on your changes?

Reply:
Yes; wrong the entire time.

Reply:
I need more explanation. What is wrong with it the way it was? Looking at `API_CALL_NOCONF` it just calls `API_SESSION_INIT` and that pretty much just sets up the session name and does a few random checks. Why is any of that bad and why should the prepare check be done before, say, checking PANIC? In fact, it seems like this new `WT_TXN_ERROR` code could be very similar to `WT_SESSION|CONN_CHECK_PANIC`. You could add a `WT_SESSION_CHECK_TXN` macro at the same place.

Reply:
The problem is that once we enter the standard API call macros, any error jumps to the `err` label, that calls `API_END`, which, if `ret != 0`, sets `WT_TXN_ERROR`. If a transaction is prepared, we aren't allowed to roll it back if the application chooses to commit it, and setting `WT_TXN_ERROR` will change the behavior of commit to roll back the function instead of commit.